File: InnovateConnect/InnovateConnect/backend/InnovateConnect.API/Controllers/AdminController.cs
using InnovateConnect.API.Data;
using InnovateConnect.API.Models;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace InnovateConnect.API.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    [Authorize(Roles = "Admin")]
    public class AdminController : ControllerBase
    {
        private readonly AppDbContext _context;

        public AdminController(AppDbContext context)
        {
            _context = context;
        }

        [HttpGet("stats")]
        public async Task<IActionResult> GetStats()
        {
            var stats = new
            {
                TotalStudents = await _context.Users.CountAsync(u => u.Role == "Student"),
                TotalCompanies = await _context.Users.CountAsync(u => u.Role == "Company"),
                TotalIdeas = await _context.Ideas.CountAsync(),
                ActiveInternships = await _context.Internships.CountAsync()
            };
            return Ok(stats);
        }

        [HttpGet("users")]
        public async Task<IActionResult> GetUsers()
        {
            var users = await _context.Users
                .Select(u => new
                {
                    u.Id,
                    u.Email,
                    u.Role,
                    u.CreatedAt,
                    ProfileName = u.Role == "Student" 
                        ? _context.Students.Where(s => s.UserId == u.Id).Select(s => s.FullName).FirstOrDefault() 
                        : u.Role == "Company" 
                            ? _context.Companies.Where(c => c.UserId == u.Id).Select(c => c.CompanyName).FirstOrDefault() 
                            : "Admin"
                })
                .ToListAsync();
            return Ok(users);
        }

        [HttpDelete("users/{id}")]
        public async Task<IActionResult> DeleteUser(int id)
        {
            var user = await _context.Users.FindAsync(id);
            if (user == null) return NotFound();

            if (user.Role == "Admin") return BadRequest("Cannot delete admin users.");

            // Manual Cascade Delete for Prototype Stability
            if (user.Role == "Student")
            {
                var student = await _context.Students.FirstOrDefaultAsync(s => s.UserId == id);
                if (student != null)
                {
                    // Delete applications
                    var apps = _context.Applications.Where(a => a.StudentId == student.Id);
                    _context.Applications.RemoveRange(apps);
                    
                    // Delete ideas
                    var ideas = _context.Ideas.Where(i => i.StudentId == student.Id);
                    _context.Ideas.RemoveRange(ideas);

                    _context.Students.Remove(student);
                }
            }
            else if (user.Role == "Company")
            {
                var company = await _context.Companies.FirstOrDefaultAsync(c => c.UserId == id);
                if (company != null)
                {
                    var internships = await _context.Internships.Where(i => i.CompanyId == company.Id).ToListAsync();
                    foreach (var i in internships)
                    {
                        var apps = _context.Applications.Where(a => a.InternshipId == i.Id);
                        _context.Applications.RemoveRange(apps);
                    }
                    _context.Internships.RemoveRange(internships);
                    _context.Companies.Remove(company);
                }
            }
            
            _context.Users.Remove(user);
            await _context.SaveChangesAsync();
            return Ok(new { Message = "User and all related data deleted successfully" });
        }

        [HttpGet("ideas")]
        public async Task<IActionResult> GetIdeas()
        {
            var ideas = await _context.Ideas
                .Include(i => i.Student)
                .Select(i => new {
                    i.Id,
                    i.Title,
                    i.Description,
                    i.TechnologyUsed,
                    i.PostedDate,
                    StudentName = i.Student != null ? i.Student.FullName : "Unknown"
                })
                .ToListAsync();
            return Ok(ideas);
        }

        [HttpDelete("ideas/{id}")]
        public async Task<IActionResult> DeleteIdea(int id)
        {
            var idea = await _context.Ideas.FindAsync(id);
            if (idea == null) return NotFound();
            _context.Ideas.Remove(idea);
            await _context.SaveChangesAsync();
            return Ok(new { Message = "Idea deleted successfully" });
        }

        [HttpGet("internships")]
        public async Task<IActionResult> GetInternships()
        {
            var internships = await _context.Internships
                .Include(i => i.Company)
                .Select(i => new {
                    i.Id,
                    i.Title,
                    i.Description,
                    i.TechnologyUsed,
                    i.Stipend,
                    i.PostedDate,
                    CompanyName = i.Company != null ? i.Company.CompanyName : "Unknown"
                })
                .ToListAsync();
            return Ok(internships);
        }

        [HttpDelete("internships/{id}")]
        public async Task<IActionResult> DeleteInternship(int id)
        {
            var internship = await _context.Internships.FindAsync(id);
            if (internship == null) return NotFound();
            
            // Delete related applications first
            var apps = _context.Applications.Where(a => a.InternshipId == id);
            _context.Applications.RemoveRange(apps);

            _context.Internships.Remove(internship);
            await _context.SaveChangesAsync();
            return Ok(new { Message = "Internship and related applications deleted successfully" });
        }

        [HttpGet("contacts")]
        public async Task<IActionResult> GetContactMessages()
        {
            var messages = await _context.ContactMessages
                .OrderByDescending(m => m.CreatedAt)
                .ToListAsync();
            return Ok(messages);
        }

        [HttpDelete("contacts/{id}")]
        public async Task<IActionResult> DeleteContactMessage(int id)
        {
            var message = await _context.ContactMessages.FindAsync(id);
            if (message == null) return NotFound();
            _context.ContactMessages.Remove(message);
            await _context.SaveChangesAsync();
            return Ok(new { Message = "Message deleted successfully" });
        }

        [HttpPut("contacts/{id}/review")]
        public async Task<IActionResult> MarkAsReviewed(int id)
        {
            var message = await _context.ContactMessages.FindAsync(id);
            if (message == null) return NotFound();

            message.IsReviewed = !message.IsReviewed;
            await _context.SaveChangesAsync();
            return Ok(new { Message = "Status updated", IsReviewed = message.IsReviewed });
        }
    }
}

File: InnovateConnectSpring/backend/src/main/java/com/innovateconnect/api/controllers/AdminController.java
package com.innovateconnect.api.controllers;

import com.innovateconnect.api.models.User;
import com.innovateconnect.api.models.ContactMessage;
import com.innovateconnect.api.repositories.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;
import lombok.extern.slf4j.Slf4j;

import org.springframework.transaction.annotation.Transactional;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/admin")
@Slf4j
public class AdminController {

    @jakarta.annotation.PostConstruct
    public void init() {
        log.info("===============================================");
        log.info("ADMIN CONTROLLER LOADED - VERSION: DEEP_DELETE_FIX_V2");
        log.info("===============================================");
    }

    @GetMapping("/version")
    public ResponseEntity<String> getVersion() {
        return ResponseEntity.ok("DEEP_DELETE_FIX_V2");
    }

    @Autowired
    private UserRepository userRepository;
    @Autowired
    private StudentRepository studentRepository;
    @Autowired
    private CompanyRepository companyRepository;
    @Autowired
    private IdeaRepository ideaRepository;
    @Autowired
    private InternshipRepository internshipRepository;
    @Autowired
    private ApplicationRepository applicationRepository;
    @Autowired
    private ContactMessageRepository contactMessageRepository;

    @GetMapping("/stats")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<?> getStats() {
        log.info("Admin fetching statistics...");
        Map<String, Long> stats = new HashMap<>();
        stats.put("totalStudents", studentRepository.count());
        stats.put("totalCompanies", companyRepository.count());
        stats.put("totalIdeas", ideaRepository.count());
        stats.put("activeInternships", internshipRepository.count());
        stats.put("pendingMessages", contactMessageRepository.countByStatus("PENDING"));
        log.info("Stats fetched: {}", stats);
        return ResponseEntity.ok(stats);
    }

    @GetMapping("/users")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<?> getAllUsers() {
        log.info("Admin fetching all users list...");
        return ResponseEntity.ok(userRepository.findAll().stream().map(u -> {
            Map<String, Object> map = new HashMap<>();
            map.put("id", u.getId());
            map.put("email", u.getEmail());
            map.put("role", u.getRole());
            map.put("createdAt", u.getCreatedAt());

            String name = u.getEmail() != null ? u.getEmail().split("@")[0] : "User";
            String role = u.getRole() != null ? u.getRole().trim() : "";
            
            if ("Student".equalsIgnoreCase(role)) {
                name = studentRepository.findByUserId(u.getId()).map(s -> s.getFullName()).orElse(name);
            } else if ("Company".equalsIgnoreCase(role)) {
                name = companyRepository.findByUserId(u.getId()).map(c -> c.getCompanyName()).orElse(name);
            }
            map.put("profileName", name);
            return map;
        }).collect(Collectors.toList()));
    }

    @DeleteMapping("/users/{id}")
    @PreAuthorize("hasRole('ADMIN')")
    @Transactional
    public ResponseEntity<?> deleteUser(@PathVariable Integer id) {
        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
        log.info("Admin [{}] attempting DEEP delete for user ID: {}", auth.getName(), id);

        User user = userRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("User not found"));
        
        if ("Admin".equalsIgnoreCase(user.getRole())) {
            return ResponseEntity.badRequest().body("Cannot delete Admin");
        }

        try {
            // Explicitly handle Student children
            studentRepository.findByUserId(id).ifPresent(student -> {
                log.info("Explicitly cleaning up Student [{}] applications and ideas...", student.getId());
                // 1. Delete all applications made by this student
                applicationRepository.deleteAll(applicationRepository.findByStudentId(student.getId()));
                // 2. Delete all ideas shared by this student
                ideaRepository.deleteAll(ideaRepository.findByStudent(student));
                // Flush to ensure deletions are recognized by DB before student profile is removed
                applicationRepository.flush(); 
                ideaRepository.flush();
            });

            // Explicitly handle Company children
            companyRepository.findByUserId(id).ifPresent(company -> {
                log.info("Explicitly cleaning up Company [{}] internships and their applications...", company.getId());
                internshipRepository.findByCompany(company).forEach(internship -> {
                    // 1. Delete all applications for this internship
                    applicationRepository.deleteAll(applicationRepository.findByInternshipId(internship.getId()));
                });
                applicationRepository.flush();
                // 2. Delete the internships themselves
                internshipRepository.deleteAll(internshipRepository.findByCompany(company));
                internshipRepository.flush();
            });

            // Finally delete the User (this will cascade to Student/Company profile via orphanRemoval)
            userRepository.delete(user);
            userRepository.flush(); 

            log.info("User ID: {} and all associated data cleared successfully.", id);
            return ResponseEntity.ok("User deleted successfully");
        } catch (Exception e) {
            log.error("CRITICAL: Error during deep delete of user ID {}: {}", id, e.getMessage());
            return ResponseEntity.status(500).body("Error deleting user: " + e.getMessage());
        }
    }

    @GetMapping("/ideas")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<?> getAllIdeas() {
        log.info("Admin fetching all ideas list...");
        return ResponseEntity.ok(ideaRepository.findAll().stream().map(idea -> {
            Map<String, Object> map = new HashMap<>();
            map.put("id", idea.getId());
            map.put("title", idea.getTitle());
            map.put("technologyUsed", idea.getTechnologyUsed());
            map.put("postedDate", idea.getPostedDate());
            map.put("studentName", idea.getStudentName());
            return map;
        }).collect(Collectors.toList()));
    }

    @DeleteMapping("/ideas/{id}")
    @PreAuthorize("hasRole('ADMIN')")
    @Transactional
    public ResponseEntity<?> deleteIdea(@PathVariable Integer id) {
        log.info("Admin deleting idea ID: {}", id);
        ideaRepository.deleteById(id);
        return ResponseEntity.ok("Idea deleted");
    }

    @GetMapping("/internships")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<?> getAllInternships() {
        log.info("Admin fetching all internships list...");
        return ResponseEntity.ok(internshipRepository.findAll().stream().map(intern -> {
            Map<String, Object> map = new HashMap<>();
            map.put("id", intern.getId());
            map.put("title", intern.getTitle());
            map.put("companyName", intern.getCompanyName());
            map.put("stipend", intern.getStipend());
            map.put("postedDate", intern.getPostedDate());
            return map;
        }).collect(Collectors.toList()));
    }

    @DeleteMapping("/internships/{id}")
    @PreAuthorize("hasRole('ADMIN')")
    @Transactional
    public ResponseEntity<?> deleteInternship(@PathVariable Integer id) {
        log.info("Admin deleting internship ID: {}", id);
        internshipRepository.deleteById(id);
        return ResponseEntity.ok("Internship deleted");
    }

    // Contact Message Management
    @GetMapping("/messages")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<?> getAllMessages() {
        log.info("Admin fetching all contact messages...");
        return ResponseEntity.ok(contactMessageRepository.findAllByOrderByCreatedAtDesc().stream().map(msg -> {
            Map<String, Object> map = new HashMap<>();
            map.put("id", msg.getId());
            map.put("name", msg.getName());
            map.put("email", msg.getEmail());
            map.put("message", msg.getMessage());
            map.put("status", msg.getStatus());
            map.put("createdAt", msg.getCreatedAt());
            return map;
        }).collect(Collectors.toList()));
    }

    @PutMapping("/messages/{id}/status")
    @PreAuthorize("hasRole('ADMIN')")
    @Transactional
    public ResponseEntity<?> updateMessageStatus(@PathVariable Integer id, @RequestBody Map<String, String> payload) {
        log.info("Admin updating message ID: {} status", id);
        return contactMessageRepository.findById(id)
            .map(msg -> {
                String newStatus = payload.get("status");
                if (newStatus != null && (newStatus.equals("PENDING") || newStatus.equals("REVIEWED"))) {
                    msg.setStatus(newStatus);
                    contactMessageRepository.save(msg);
                    return ResponseEntity.ok("Message status updated");
                }
                return ResponseEntity.badRequest().body("Invalid status");
            })
            .orElse(ResponseEntity.notFound().build());
    }

    @DeleteMapping("/messages/{id}")
    @PreAuthorize("hasRole('ADMIN')")
    @Transactional
    public ResponseEntity<?> deleteMessage(@PathVariable Integer id) {
        log.info("Admin deleting message ID: {}", id);
        contactMessageRepository.deleteById(id);
        return ResponseEntity.ok("Message deleted");
    }
}

